shared/js/background/components/tab-tracking.js

```diff
@@ -0,0 +1,119 @@
+/* global BUILD_TARGET */
+import browser from 'webextension-polyfill'
+import { restoreDefaultClickToLoadRuleActions } from '../dnr-click-to-load'
+import Companies from '../companies'
+import { isRedirect } from '../utils'
+import * as devtools from '../devtools'
+
+/**
+ * @typedef {import('../tab-manager.js')} TabManager
+ * @typedef {import('../devtools')} Devtools
+ */
+
+export default class TabTracker {
+    /**
+     * @param {{
+    *  tabManager: TabManager;
+    * }} options
+    */
+    constructor ({ tabManager }) {
+        this.tabManager = tabManager
+        this.createdTargets = new Map()
+
+        browser.webRequest.onHeadersReceived.addListener((request) => {
```

Comment:
Are we moving away from the pattern of putting all extension event listeners in events.js? Fine by me if so, but perhaps we should make that clear by adding a comment in events.js or something?

Reply:
Is there a notable cost to multiple listeners? I've always avoided it; perhaps prematurely.

Either way we would now have more places we're binding to the extension APIs, would it make more sense for events.js to fire events that components can listen to?

Reply:
Yeah, I'd prefer to put all the code for a feature together in a single file, rather than having events in `event.js`, messages in `message-handlers.js` and so on. This makes it easier to track the surface area of each feature.

Regarding registering event handlers in multiple places - my assumption is that this shouldn't degrade performance, except perhaps in the sync case that Dave mentioned. It would be surprising if this assumption was incorrect.